### PR TITLE
gh-96761: Fix lto-build for clang

### PR DESCRIPTION
**Pull request by matthiasgoergens**
When enabling link time optimization for clang, you eventually get the following error:

`Modules/getbuildinfo.o: file not recognized: file format not recognized`

This patch fixes that problem.

* Issue: gh-96761

### CODE REVIEW

**Review by corona10 — CHANGES_REQUESTED** [review]
Please re-generate configure file by using the following command.

````
docker run --rm --pull=always -v$(pwd):/src quay.io/tiran/cpython_autoconf:269
````

File: configure.ac [hunk 1]
```diff
@@ -1829,7 +1829,7 @@ if test "$Py_LTO" = 'true' ; then
   case $CC in
     *clang*)
       dnl flag to disable lto during linking
-      LDFLAGS_NOLTO="-fno-lto"
+      LDFLAGS_NOLTO="-flto=${Py_LTO_POLICY}"
```
Comment (tiran) [inline configure.ac]: `LDFLAGS_NOLTO` should be linker flags without LTO, hence the name `no lto`.

Comment (matthiasgoergens) [reply configure.ac]: I guess I'll need to make a bigger change, that changes the name to something more descriptive.

```
# LDFLAGS_NOLTO is an extra flag to disable lto. It is used to speed up building
# of _bootstrap_python and _freeze_module tools, which don't need LTO.
```

For clang that seems to work better when they are the same as for the rest of the build.  The thin-lto build is already plenty fast (and even then, a slow build is better than a broken one).

**Comment by matthiasgoergens** [comment]
> Please re-generate configure file by using the following command.
> 
> ```
> docker run --rm --pull=always -v$(pwd):/src quay.io/tiran/cpython_autoconf:269
> ```

Done!

File: configure.ac (hunk 1 above)
Comment (tiran) [reply configure.ac]: clang does not support linking with objects that have LTO information? It works fine for GCC. Could you please update the comment on line 1831?

**Comment by matthiasgoergens** [comment]
I have made the requested changes; please review again.

File: configure.ac (hunk 1 above)
Comment (matthiasgoergens) [reply configure.ac]: I am not quite sure what the exact rootcause is.  I just noticed that when I tried to build some things with LTO and some things (like `_bootstrap_python`) without LTO, I get an error from the linker via clang.

Comment (matthiasgoergens) [reply configure.ac]: I updated the comment.

Comment (tiran) [reply configure.ac]: Does thin LTO linking work with object files that are built with full LTO? This would speed up LTO builds a bit.

Comment (corona10) [reply configure.ac]: IIUC, at this moment ``-flto=${Py_LTO_POLICY}`` could be ``-flto=default`` so it's wrong.

**Review by corona10 — CHANGES_REQUESTED** [review]
see: https://github.com/python/cpython/pull/96762#discussion_r968204126

File: configure.ac (hunk 1 above)
Comment (matthiasgoergens) [reply configure.ac]: Hmm, let me think about how to fix this.  Or do you have a suggestion?

File: configure.ac [hunk 2]
```diff
@@ -1828,8 +1828,9 @@ esac
 if test "$Py_LTO" = 'true' ; then
   case $CC in
     *clang*)
-      dnl flag to disable lto during linking
-      LDFLAGS_NOLTO="-fno-lto"
+      dnl For clang, it seems our LTO policy for bootstrap and freezing needs
+      dnl to match the rest of the build.
+      LDFLAGS_NOLTO="-flto=${Py_LTO_POLICY}"
```
Comment (tiran) [inline configure.ac]:
```suggestion
      dnl Clang linker requires -flto in order to link objects with LTO information.
      dnl Thin LTO is faster and works for object files with full LTO information, too.
      LDFLAGS_NOLTO="-flto=thin"
```

Comment (corona10) [reply configure.ac]: Old clang does not support ThinLTO.
(https://releases.llvm.org/4.0.0/tools/clang/docs/ThinLTO.html, Clang 4 was released on 13th March 2017)
Please take a look at my PR: https://github.com/python/cpython/pull/96766

**Comment by matthiasgoergens** [comment]
@corona10 Just to be sure: does your PR at https://github.com/python/cpython/pull/96766 obsolete this one?

If yes, I'll close mine.

**Comment by corona10** [comment]
@matthiasgoergens cc @tiran 
There are several consensuses on this issue.
A. Using ThinLTO as the default LTO policy for clang is not decided yet, maybe #96766 will cover the issue.
B. If #96766 is merged we can use ThinLTO for available clang for `bootstrap_python` also if needed.
C. But A and B do not mean that `bootstrap_python` should be built with `LTO`on clang. (If need we should use it but if not we don't have to use it) since I can not meet the same issue on macOS with clang so I need to reproduce the issue on Linux first. I need to know the root cause.

**Comment by corona10** [comment]
> @corona10 Just to be sure: does your PR at https://github.com/python/cpython/pull/96766 obsolete this one?

So in conclusion, it's a slightly different issue.

**Comment by dwvshell** [comment]
How do I put this to work

File: configure.ac [hunk 3]
```diff
@@ -1828,8 +1828,9 @@ esac
 if test "$Py_LTO" = 'true' ; then
   case $CC in
     *clang*)
-      dnl flag to disable lto during linking
-      LDFLAGS_NOLTO="-fno-lto"
+      dnl Clang linker requires -flto in order to link objects with LTO information.
+      dnl Thin LTO is faster and works for object files with full LTO information, too.
+      LDFLAGS_NOLTO="-flto=thin"
```
Comment (corona10) [inline configure.ac]:
```suggestion
      AX_CHECK_COMPILE_FLAG([-flto=thin],[LDFLAGS_NOLTO="-flto=thin"],[LDFLAGS_NOLTO="-flto"])
```

Comment (corona10) [inline configure.ac:1833]:
```suggestion
      LDFLAGS_NOLTO=""
```

**Review by corona10 — CHANGES_REQUESTED** [review]
@matthiasgoergens 

* Please rebase the PR and regen configure with the suggested changes.
* Please update the following comment also (for clang case).
https://github.com/python/cpython/blob/1fc8bd3710670982bc7ce91a34d8cd4bcdf88b9a/Makefile.pre.in#L94-L95

**Review by corona10 — COMMENTED** [review]
Would you like to regenerate the configure by using `docker run --rm --pull=always -v$(pwd):/src quay.io/tiran/cpython_autoconf:269`?

**Comment by matthiasgoergens** [comment]
Closed in favour of https://github.com/python/cpython/pull/96945

**Pull request closed without merging**